pkg/controller/v1alpha1/inferencegraph/controller.go

```diff
@@ -0,0 +1,266 @@
+	"knative.dev/pkg/kmp"
+	"knative.dev/serving/pkg/apis/autoscaling"
+	knservingv1 "knative.dev/serving/pkg/apis/serving/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+// InferenceGraphReconciler reconciles a InferenceGraph object
+type InferenceGraphReconciler struct {
+	client.Client
+	Log      logr.Logger
+	Scheme   *runtime.Scheme
+	Recorder record.EventRecorder
+}
+
+// InferenceGraphState describes the Readiness of the InferenceGraph
+type InferenceGraphState string
+
+const (
+	InferenceGraphNotReadyState InferenceGraphState = "InferenceGraphNotReady"
+	InferenceGraphReadyState    InferenceGraphState = "InferenceGraphReady"
+)
+
+type RouterConfig struct {
+	Image         string `json:"image"`
+	CpuRequest    string `json:"cpuRequest"`
+	CpuLimit      string `json:"cpuLimit"`
+	MemoryRequest string `json:"memoryRequest"`
+	MemoryLimit   string `json:"memoryLimit"`
+}
+
+func getRouterConfigs(configMap *v1.ConfigMap) (*RouterConfig, error) {
+
+	routerConfig := &RouterConfig{}
+	if agentConfigValue, ok := configMap.Data["router"]; ok {
+		err := json.Unmarshal([]byte(agentConfigValue), &routerConfig)
+		if err != nil {
+			panic(fmt.Errorf("Unable to unmarshall agent json string due to %v ", err))
+		}
+	}
+
+	//Ensure that we set proper values for CPU/Memory Limit/Request
+	resourceDefaults := []string{routerConfig.MemoryRequest,
+		routerConfig.MemoryLimit,
+		routerConfig.CpuRequest,
+		routerConfig.CpuLimit}
+	for _, key := range resourceDefaults {
+		_, err := resource.ParseQuantity(key)
+		if err != nil {
+			return routerConfig, fmt.Errorf("Failed to parse resource configuration for router: %q",
+				err.Error())
+		}
+	}
+
+	return routerConfig, nil
+}
+
+func (r *InferenceGraphReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	_ = context.Background()
+
+	// Fetch the InferenceService instance
+	graph := &v1alpha1api.InferenceGraph{}
+	if err := r.Get(ctx, req.NamespacedName, graph); err != nil {
+		if apierr.IsNotFound(err) {
+			// Object not found, return.  Created objects are automatically garbage collected.
+			// For additional cleanup logic use finalizers.
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+
+	r.Log.Info("Reconciling inference graph", "apiVersion", graph.APIVersion, "graph", graph.Name)
+	configMap := &v1.ConfigMap{}
+	err := r.Client.Get(context.TODO(), types.NamespacedName{Name: constants.InferenceServiceConfigMapName, Namespace: constants.KServeNamespace}, configMap)
+	if err != nil {
+		r.Log.Error(err, "Failed to find config map", "name", constants.InferenceServiceConfigMapName)
+		return reconcile.Result{}, err
+	}
+	routerConfig, err := getRouterConfigs(configMap)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	// Create service if does not exist
+	desired := createKnativeService(graph.ObjectMeta, graph, routerConfig)
```

Comment:
Before we already dropped the hard requirement for knative by deployment and HPA, so for graph do we consider add adaptation to this? Or I think this will add back the dependency on knative.

Reply:
We will still support the both Knative and raw deployment mode, will add the raw deployment handling as well.

Reply:
Yeah, great!